### PR TITLE
Refactor/16226/auth screens components migration to functional

### PR DESCRIPTION
**Pull request by barttom**
### Details

Migration component from lass to functional
### Fixed Issues

$ https://github.com/Expensify/App/issues/16226
PROPOSAL: https://github.com/Expensify/App/issues/16226

### Tests

1. Open the app and log in.
2. Navigate to a few screens to make sure that navigator works properly.
3. Logout and login again.
- [x] Verify that no errors appear in the JS console

### QA Steps

1. Open the app and log in.
2. Navigate to a few screens to make sure that navigator works properly.
3. Logout and login again.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
  - [x] A similar component doesn't exist in the codebase
- [x] All props are defined accurately and each prop has a `/** comment above it */`
- [x] The file is named correctly
- [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
- [x] The only data being stored in the state is data necessary for rendering and nothing else
- [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
- [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
- [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
- [x] All JSX used for rendering exists in the render method
- [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
  - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
  - [x] A similar style doesn't already exist
- [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

</details>

<details>
<summary>Mobile Web - Chrome</summary>

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/7682108/6e0716fd-c6a5-499c-b578-ab00b128cbcd

</details>

<details>
<summary>Desktop</summary>

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/7682108/1ad3b429-83ae-4726-beea-1308194b09b0

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/7682108/50e31817-fb5d-47d3-b6ef-e4157a4a146b

</details>

### CODE REVIEW

File: src/libs/Navigation/AppNavigator/AuthScreens.js
Comment (fabioh8010) [inline src/libs/Navigation/AppNavigator/AuthScreens.js:298]: @barttom You can use `useWindowDimensions` hook instead of the HOC

File: src/libs/Navigation/AppNavigator/AuthScreens.js
Comment (gedu) [inline src/libs/Navigation/AppNavigator/AuthScreens.js:133]: I think this should be inside of a `useEffect`, because `useEffect` is called when the render finished, in this case you are just checking the first time the code runs but not when the UI is done

Comment (barttom) [reply src/libs/Navigation/AppNavigator/AuthScreens.js:133]: This is what I want to achieve.
In the class implementation, `Timing` is started in the constructor and stopped in componentDidMount.
https://github.com/Expensify/App/blob/main/src/libs/Navigation/AppNavigator/AuthScreens.js

**Comment by situchan** [comment]
I can review if this requires C+

**Comment by marcochavezf** [comment]
Hi @situchan, yes please, I will assign you as C+ reviewer

**Comment by marcochavezf** [comment]
Hi @barttom, there are some lint errors and unverified commits. For the unverified commits, I think the solution will be maybe to force push the branch with verified commits.

File: src/libs/Navigation/AppNavigator/AuthScreens.js
Comment (situchan) [inline src/libs/Navigation/AppNavigator/AuthScreens.js:222]: Please fix lint

Screenshot 2023-09-25 at 11 33 08 PM

**Review by situchan — COMMENTED** [review]
Code looks good. Let's just fix lint

**Comment by barttom** [comment]
@situchan fixed

**Comment by barttom** [comment]
@situchan fixed

File: src/libs/Navigation/AppNavigator/AuthScreens.js
Comment (situchan) [inline src/libs/Navigation/AppNavigator/AuthScreens.js:216]: This is root component and it's important to prevent unnecessary re-rendering.
So let's use `memo` which is equivalent to `shouldComponentUpdate` in function component

Comment (barttom) [reply src/libs/Navigation/AppNavigator/AuthScreens.js:216]: In function components we should use hooks instead of HOCs, so there is no information about windowHeight & isSmallScreenWidth in props. I've checked re-rendering and we don't have regression in this field. Please notice, that the only one dependency we have is `isSmallScreenWidth` and the component will re-render only if that variable has been changed.
This optimization is no longer needed :)

Comment (situchan) [reply src/libs/Navigation/AppNavigator/AuthScreens.js:216]: > so there is no information about windowHeight & isSmallScreenWidth in props

We can access in `memo`
i.e.
https://github.com/Expensify/App/blob/d4a33cd5a2f21d2ee853682f8938a965b43c52b9/src/components/PopoverWithMeasuredContent.js#L164-L169

Comment (situchan) [reply src/libs/Navigation/AppNavigator/AuthScreens.js:216]: @barttom bump ^

We don't need to re-render this component when seesion, lastOpenedPublicRoomID, isUsingMemoryOnlyKeys, lastUpdateIDAppliedToClient Onyx prop values are changed as they're used only on component mount.

Comment (barttom) [reply src/libs/Navigation/AppNavigator/AuthScreens.js:216]: @situchan sorry for not responding. I’m OOO until end of this week too.

I used hook instead for f HOC so I don’t have access to the dimensions props.
Using hooks is more optimal and previous shouldComponentUpdate is not needed anymore

Comment (situchan) [reply src/libs/Navigation/AppNavigator/AuthScreens.js:216]: I am not suggesting HOC. I am not sure what you're talking about.

Comment (barttom) [reply src/libs/Navigation/AppNavigator/AuthScreens.js:216]: On first comment You suggest using memo and rewrite previous implementation of componentShouldUpdate. Without HOC we don’t have dimensions props anymore.

Comment (situchan) [reply src/libs/Navigation/AppNavigator/AuthScreens.js:216]: Of course, if using `useDimensions`, component is always re-rendered. So we can simply return false in memo function so that other props changes don't cause component re-render.

Comment (barttom) [reply src/libs/Navigation/AppNavigator/AuthScreens.js:219]: @situchan memoization added

**Comment by situchan** [comment]
Please add test case

**Comment by barttom** [comment]
> Please add test case

test case added

**Comment by barttom** [comment]
@situchan could You review this?

**Comment by situchan** [comment]
reviewing today

**Comment by barttom** [comment]
@situchan can we move it forward?

**Comment by marcochavezf** [comment]
friendly bump @situchan

**Comment by situchan** [comment]
@barttom add logout -> relogin flow in Tests.
Also, please update QA Steps so that QA team can test.  It's fine to add like this "Same as Tests Steps"

**Comment by barttom** [comment]
@situchan done
Is everything fine now in the description? Please add all comments at once to speed up this one.

**Comment by barttom** [comment]
@situchan friendly bump :)

**Comment by situchan** [comment]
Thanks for the bump. I will complete review today

**Comment by situchan** [comment]
@barttom please apply changes from #29063. Ctrl + J not working

**Comment by barttom** [comment]
@situchan thanks for catching this! Fixed ;)

**Comment by situchan** [comment]
Sorry but let's hold this 1-2 days until #29323 is merged.
I checked through open PRs and it's the only one recently touched AuthScreens.js and likely to be merged.

**Comment by barttom** [comment]
OK. Can You just confirm it it works properly? I want to fix eventual errors in the meantime.

**Comment by situchan** [comment]
@barttom that PR was merged. Please pull main

**Comment by barttom** [comment]
@situchan done :)

**Comment by situchan** [comment]
Crash on resize window on web. Already happening on main so out of scope.


https://github.com/Expensify/App/assets/108292595/38226ae0-9047-4bb6-aefe-3d253e0b7223

**Comment by situchan** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>


https://github.com/Expensify/App/assets/108292595/5999df00-4021-4764-b6a4-01971943de07



</details>

<details>
<summary>Mobile Web - Chrome</summary>


https://github.com/Expensify/App/assets/108292595/381cf3e7-64ee-4790-8b1c-f2ff6fac1fce



</details>

<details>
<summary>Mobile Web - Safari</summary>


https://github.com/Expensify/App/assets/108292595/fc5abf1c-727d-4bf7-9ef7-a73d89c7f8a0



</details>

<details>
<summary>Desktop</summary>


https://github.com/Expensify/App/assets/108292595/2add6197-780b-4a2c-bd76-d920acc53b46



</details>

<details>
<summary>iOS</summary>


https://github.com/Expensify/App/assets/108292595/522ffb2c-f550-4056-ad4b-d51c6b5a8b6b



</details>

<details>
<summary>Android</summary>


https://github.com/Expensify/App/assets/108292595/c97562eb-2a65-4efb-b8c0-bcd363dfcc0f



</details>

**Review by situchan — APPROVED** [review]
:shipit:

**Review by marcochavezf — APPROVED** [review]
Code LGTM, thank you guys!

File: src/libs/Navigation/AppNavigator/AuthScreens.js
Comment (situchan) [inline src/libs/Navigation/AppNavigator/AuthScreens.js:173]: As we are aware already, there's hidden bug where `lastUpdateIDAppliedToClient` is NOT the updated updateID we saved in Onyx as this isn't added in useEffect dependency

**Pull request merged**
